### DISCUSSION TITLE
Cross-platform tasks? Pehaps via bundled Python scripts?

### DISCUSSION THREAD

**Discussion by aaronsteers** [copier-org/copier#649] [Q&A] [answered]
I am evaluating Copier as a means of installing resources for another open source project. One of our use cases is to inject entries into a yaml file if the file exists, or create a new file otherwise.

Tasks seem to be a great way to have this more powerful customized install logic, but by nature of being tied to terminal syntax, it seems these would necessarily be OS-specific. How are folks working around this today, for instance, if a template needs to be supported on Windows, Mac, and Linux?

My first thought was to bundle a python script that could perform these actions using platform-agnostic core python libraries.

Questions:

- Are Python script tasks an approach that has some precedent for `tasks` and `migrations`?
- Any way to bundle the scripts for task execution and migration but not install them into the users directory? (Perhaps deleting after executing?)

Curious what strategies have worked for other folks. Thanks!

**Comment by pawamoy** [comment, marked as answer]
You can definitely add a task that runs a Python script. This Python script can then delete itself indeed:

```yaml
_tasks:
- "{{ _copier_python }} postgen.py"
```

```python
# postgen.py, at the root of your template
def update_yaml_file():
    ...

def delete_self():
    os.remove(__file__)

def main():
    update_yaml_file()
    delete_self()

if __name__ == "__main__":
    main()
```

Note that `{{ _copier_python }}` is the Python interpreter used by Copier, so if Copier was installed in an isolated venv (through pipx for example), then you will only have access to the packages installed in this venv. If you are *absolutely sure* that there will always be a `python` or `python3` (or `/usr/local/bin/python` or Anaconda, etc.) executable available in PATH on the target machines, then you can use just that.

**Reply by yajo** [reply]
You can execute scripts in the template folder, so you don't need to self-delete them. Example: https://github.com/Tecnativa/doodba-copier-template/blob/11e31c188084449f99dca38de739a4b25fe0c9db/copier.yml#L51

**Reply by bittner** [reply]
It would help to have this information in the Copier docs. Maybe even replace the shell scripting examples and favor this as a system-independent approach.
